lib/api/CSerializableToJson.cc

```diff
@@ -7,75 +7,139 @@
+CSerializableToCompressedChunkedJson::CSerializableToCompressedChunkedJson(std::size_t maxDocumentSize)
+    : m_MaxDocumentSize{std::min(maxDocumentSize, MAX_DOCUMENT_SIZE)} {
+}
+
+std::stringstream CSerializableToCompressedChunkedJson::jsonCompressedStream() const {
+    std::stringstream result;
+    compressAndEncode(this->callableAddToJsonStream(), result);
+    return result;
+}
+
+void CSerializableToCompressedChunkedJson::addCompressedToJsonStream(
     const std::string& compressedDocTag,
-    const std::string& payloadTag) const {
-    std::stringstream compressedStream{this->jsonCompressedStream()};
-    std::streamsize processed{0};
-    compressedStream.seekg(0, compressedStream.end);
-    std::streamsize remained{compressedStream.tellg()};
-    compressedStream.seekg(0, compressedStream.beg);
+    const std::string& payloadTag,
+    TRapidJsonWriter& writer) const {
+
+    using TCharVec = std::vector<char>;
+
+    TCharVec buffer;
+    buffer.reserve(m_MaxDocumentSize);
+    io::stream<io::back_insert_device<TCharVec>> output{io::back_inserter(buffer)};
+    compressAndEncode(this->callableAddToJsonStream(), output);
+
     std::size_t docNum{0};
-    std::string buffer;
-    while (remained > 0) {
-        std::size_t bytesToProcess{std::min(MAX_DOCUMENT_SIZE, static_cast<size_t>(remained))};
-        buffer.clear();
-        std::copy_n(std::istreambuf_iterator<char>(compressedStream.seekg(processed)),
-                    bytesToProcess, std::back_inserter(buffer));
-        remained -= bytesToProcess;
-        processed += bytesToProcess;
+    for (std::size_t i = 0; i < buffer.size(); i += m_MaxDocumentSize) {
+        rapidjson::SizeType bytesToWrite{static_cast<rapidjson::SizeType>(
+            std::min(m_MaxDocumentSize, buffer.size() - i))};
+
         writer.StartObject();
         writer.Key(compressedDocTag);
         writer.StartObject();
         writer.Key(JSON_DOC_NUM_TAG);
         writer.Uint64(docNum);
         writer.Key(payloadTag);
-        writer.String(buffer);
-        if (remained == 0) {
+        writer.String(&buffer[i], bytesToWrite);
+        if (i + bytesToWrite == buffer.size()) {
             writer.Key(JSON_EOS_TAG);
             writer.Bool(true);
         }
         writer.EndObject();
         writer.EndObject();
+
         ++docNum;
     }
 }
+
+CSerializableFromCompressedChunkedJson::TIStreamPtr
+CSerializableFromCompressedChunkedJson::rawJsonStream(const std::string& compressedDocTag,
+                                                      const std::string& payloadTag,
+                                                      TIStreamPtr inputStream,
+                                                      std::iostream& buffer) {
+    if (inputStream != nullptr) {
+        rapidjson::IStreamWrapper isw{*inputStream};
+        try {
+            rapidjson::Document doc;
+            bool done{false};
+            do {
+                doc.ParseStream<rapidjson::kParseStopWhenDoneFlag>(isw);
```

Comment:
I think we need to check for a parse error after this as the RapidJSON way of doing things is to return a doc that contains the error information.  `ifExists` might prevent undefined behaviour in this case but it will lose the more useful diagnostic information of what was wrong.  There may well be other places in our code where we should be doing this.

```suggestion
                doc.ParseStream<rapidjson::kParseStopWhenDoneFlag>(isw);
                if (doc.HasParseError()) {
                    const char* error{rapidjson::GetParseError_En(doc.GetParseError())};
                    throw std::runtime_exception{"Error parsing JSON at offset " + std::to_string(doc.GetErrorOffset()) + ": " + ((error != nullptr) ? error : "No message")};
                }
```

Reply:
Good suggestion! This behaviour was just copied over from the original implementation. There are probably a couple of other places we should be doing the same.

Reply:
This was done in [65348db](https://github.com/elastic/ml-cpp/pull/1897/commits/65348dbe39335647335e2093a39f2fe0e8fdfca5).